In [477]:
import pandas as pd
import numpy as np
import sqlite3

In [442]:
# Global E-Commerce dataset
ecommerce_df = pd.read_csv("../Data/global_ecommerce_sales.csv")

# E-Commerce Sales Transactions dataset
transaction_df = pd.read_csv("../Data/ecommerce_sales_transactions.csv")

In [443]:
ecommerce_df.head()

,Order_ID,Order_Date,Customer_Name,Customer_Segment,Country,Region,Product_Category,Product_Name,Quantity,Unit_Price,Discount_Percent,Total_Sales,Shipping_Cost,Profit,Payment_Method
0,ORD-11121,2023-01-02,Karen Suzuki,Corporate,United States,North America,Technology,Wireless Bluetooth Headphones,3,99.43,0,298.29,9.31,124.92,Cash on Delivery
1,ORD-11244,2023-01-02,John Johansson,Corporate,Spain,Europe,Technology,Mechanical Gaming Keyboard,4,97.93,20,313.38,14.31,83.62,Cash on Delivery
2,ORD-10325,2023-01-03,Jessica Garcia,Consumer,Mexico,North America,Office Supplies,Binder Clips Assorted 48pc,2,10.74,0,21.48,8.12,3.69,Credit Card
3,ORD-10467,2023-01-03,Clara Taylor,Corporate,Italy,Europe,Technology,Webcam HD 1080p,2,61.86,15,105.16,10.79,26.32,PayPal
4,ORD-11454,2023-01-05,Felix Thomas,Consumer,Italy,Europe,Furniture,Standing Desk Converter,4,330.67,20,1058.14,11.09,253.44,Credit Card


In [444]:
transaction_df.head()

,order_id,customer_id,product_id,category,price,discount,quantity,payment_method,order_date,delivery_time_days,region,returned,total_amount,shipping_cost,profit_margin,customer_age,customer_gender
0,O100000,C17270,P234890,Home,164.08,0.15,1,Credit Card,2023-12-23,4,West,No,139.47,7.88,31.17,60,Female
1,O100001,C17603,P228204,Grocery,24.73,0.00,1,Credit Card,2025-04-03,6,South,No,24.73,4.60,-2.62,37,Male
2,O100002,C10860,P213892,Electronics,175.58,0.05,1,Credit Card,2024-10-08,4,North,No,166.80,6.58,13.44,34,Male
3,O100003,C15390,P208689,Electronics,63.67,0.00,1,UPI,2024-09-14,6,South,No,63.67,5.50,2.14,21,Female
4,O100004,C15226,P228063,Home,16.33,0.15,1,COD,2024-12-21,6,East,No,13.88,2.74,1.15,39,Male


In [445]:
print("ECommerce:", ecommerce_df.shape)
print("Transaction:", transaction_df.shape)

ECommerce: (2000, 15)
Transaction: (34500, 17)


In [446]:
print(ecommerce_df.columns.tolist())

['Order_ID', 'Order_Date', 'Customer_Name', 'Customer_Segment', 'Country', 'Region', 'Product_Category', 'Product_Name', 'Quantity', 'Unit_Price', 'Discount_Percent', 'Total_Sales', 'Shipping_Cost', 'Profit', 'Payment_Method']


In [447]:
print(transaction_df.columns.tolist())

['order_id', 'customer_id', 'product_id', 'category', 'price', 'discount', 'quantity', 'payment_method', 'order_date', 'delivery_time_days', 'region', 'returned', 'total_amount', 'shipping_cost', 'profit_margin', 'customer_age', 'customer_gender']


In [448]:
ecommerce_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order_ID          2000 non-null   str    
 1   Order_Date        2000 non-null   str    
 2   Customer_Name     2000 non-null   str    
 3   Customer_Segment  2000 non-null   str    
 4   Country           2000 non-null   str    
 5   Region            2000 non-null   str    
 6   Product_Category  2000 non-null   str    
 7   Product_Name      2000 non-null   str    
 8   Quantity          2000 non-null   int64  
 9   Unit_Price        2000 non-null   float64
 10  Discount_Percent  2000 non-null   int64  
 11  Total_Sales       2000 non-null   float64
 12  Shipping_Cost     2000 non-null   float64
 13  Profit            2000 non-null   float64
 14  Payment_Method    2000 non-null   str    
dtypes: float64(4), int64(2), str(9)
memory usage: 234.5 KB


In [449]:
transaction_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 34500 entries, 0 to 34499
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            34500 non-null  str    
 1   customer_id         34500 non-null  str    
 2   product_id          34500 non-null  str    
 3   category            34500 non-null  str    
 4   price               34500 non-null  float64
 5   discount            34500 non-null  float64
 6   quantity            34500 non-null  int64  
 7   payment_method      34500 non-null  str    
 8   order_date          34500 non-null  str    
 9   delivery_time_days  34500 non-null  int64  
 10  region              34500 non-null  str    
 11  returned            34500 non-null  str    
 12  total_amount        34500 non-null  float64
 13  shipping_cost       34500 non-null  float64
 14  profit_margin       34500 non-null  float64
 15  customer_age        34500 non-null  int64  
 16  customer_gender

In [450]:
ecommerce_df.isnull().sum()

Order_ID            0
Order_Date          0
Customer_Name       0
Customer_Segment    0
Country             0
Region              0
Product_Category    0
Product_Name        0
Quantity            0
Unit_Price          0
Discount_Percent    0
Total_Sales         0
Shipping_Cost       0
Profit              0
Payment_Method      0
dtype: int64

In [451]:
transaction_df.isnull().sum()

order_id              0
customer_id           0
product_id            0
category              0
price                 0
discount              0
quantity              0
payment_method        0
order_date            0
delivery_time_days    0
region                0
returned              0
total_amount          0
shipping_cost         0
profit_margin         0
customer_age          0
customer_gender       0
dtype: int64

In [452]:
print("ECommerce:", ecommerce_df.duplicated().sum())
print("Transaction:", transaction_df.duplicated().sum())

ECommerce: 0
Transaction: 0


In [453]:
comparison = pd.DataFrame({
    "ECommerce": pd.Series(ecommerce_df.columns),
    "Transaction": pd.Series(transaction_df.columns)
})

comparison

,ECommerce,Transaction
0,Order_ID,order_id
1,Order_Date,customer_id
2,Customer_Name,product_id
3,Customer_Segment,category
4,Country,price
5,Region,discount
6,Product_Category,quantity
7,Product_Name,payment_method
8,Quantity,order_date
9,Unit_Price,delivery_time_days


# Data Cleaning and Standardization

### Clean Column Names
Standardize column names by converting them to lowercase,
removing extra spaces, and replacing spaces with underscores.

### Clean column names with manual code
#ecommerce_df.columns = (ecommerce_df.columns.str.lower().str.strip().str.replace(" ", "_"))
#transaction_df.columns = (transaction_df.columns.str.lower().str.strip().str.replace(" ", "_"))

In [454]:
# Function to clean column names 
def clean_column_names(df):
    df.columns = (df.columns.str.lower().str.strip().str.replace(" ", "_")
    )
    return df

ecommerce_df_df = clean_column_names(ecommerce_df)
transaction_df = clean_column_names(transaction_df)

# Display updated columns
print(ecommerce_df.columns)
print(transaction_df.columns)

Index(['order_id', 'order_date', 'customer_name', 'customer_segment',
       'country', 'region', 'product_category', 'product_name', 'quantity',
       'unit_price', 'discount_percent', 'total_sales', 'shipping_cost',
       'profit', 'payment_method'],
      dtype='str')
Index(['order_id', 'customer_id', 'product_id', 'category', 'price',
       'discount', 'quantity', 'payment_method', 'order_date',
       'delivery_time_days', 'region', 'returned', 'total_amount',
       'shipping_cost', 'profit_margin', 'customer_age', 'customer_gender'],
      dtype='str')


In [455]:
print("Global Dataset")
ecommerce_df.info()

print("\nTransaction Dataset")
transaction_df.info()

Global Dataset
<class 'pandas.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          2000 non-null   str    
 1   order_date        2000 non-null   str    
 2   customer_name     2000 non-null   str    
 3   customer_segment  2000 non-null   str    
 4   country           2000 non-null   str    
 5   region            2000 non-null   str    
 6   product_category  2000 non-null   str    
 7   product_name      2000 non-null   str    
 8   quantity          2000 non-null   int64  
 9   unit_price        2000 non-null   float64
 10  discount_percent  2000 non-null   int64  
 11  total_sales       2000 non-null   float64
 12  shipping_cost     2000 non-null   float64
 13  profit            2000 non-null   float64
 14  payment_method    2000 non-null   str    
dtypes: float64(4), int64(2), str(9)
memory usage: 234.5 KB

Transaction Dataset
<class 'p

### Standardize column names 

To create a consistent database schema, the column names in the Global E-Commerce dataset were standardized to match those in the E-Commerce Sales Transactions dataset. This makes it easier to work with both datasets when creating the SQLite database and writing SQL queries.

In [456]:
ecommerce_df = ecommerce_df.rename(columns={
    "unit_price": "price",
    "discount_percent": "discount",
    "total_sales": "total_amount",
    "product_category": "category"
})


In [457]:
# Convert order_date to datetime
ecommerce_df["order_date"] = pd.to_datetime(ecommerce_df["order_date"])
transaction_df["order_date"] = pd.to_datetime(transaction_df["order_date"])

# Verify the changes
print(ecommerce_df["order_date"].dtype)
print(transaction_df["order_date"].dtype)

datetime64[us]
datetime64[us]


In [458]:
# Check for missing values 
print("Global Ecommerce Dataset")
print(ecommerce_df.isnull().sum())

print("\nTransaction Dataset")
print(transaction_df.isnull().sum())

Global Ecommerce Dataset
order_id            0
order_date          0
customer_name       0
customer_segment    0
country             0
region              0
category            0
product_name        0
quantity            0
price               0
discount            0
total_amount        0
shipping_cost       0
profit              0
payment_method      0
dtype: int64

Transaction Dataset
order_id              0
customer_id           0
product_id            0
category              0
price                 0
discount              0
quantity              0
payment_method        0
order_date            0
delivery_time_days    0
region                0
returned              0
total_amount          0
shipping_cost         0
profit_margin         0
customer_age          0
customer_gender       0
dtype: int64


In [459]:
# Check for duplicate rows 
print("Ecommerce Dataset:", ecommerce_df.duplicated().sum())

print("Transaction Dataset:", transaction_df.duplicated().sum())

Ecommerce Dataset: 0
Transaction Dataset: 0


In [460]:
ecommerce_df.describe()

transaction_df.describe()

,price,discount,quantity,order_date,delivery_time_days,total_amount,shipping_cost,profit_margin,customer_age
count,34500.000000,34500.000000,34500.000000,34500,34500.000000,34500.000000,34500.000000,34500.000000,34500.000000
mean,119.391632,0.049291,1.490725,2024-09-13 13:39:17.843478,4.814203,170.008494,6.152120,28.116505,43.474377
min,1.010000,0.000000,1.000000,2023-09-12 00:00:00,3.000000,0.820000,0.000000,-6.200000,18.000000
25%,16.690000,0.000000,1.000000,2024-03-14 00:00:00,4.000000,19.710000,4.420000,1.500000,31.000000
50%,45.660000,0.000000,1.000000,2024-09-16 00:00:00,5.000000,56.820000,6.090000,10.550000,43.000000
75%,130.950000,0.100000,2.000000,2025-03-17 00:00:00,6.000000,168.530000,7.830000,33.132500,56.000000
max,2930.470000,0.300000,5.000000,2025-09-11 00:00:00,13.000000,12931.800000,15.650000,1536.170000,69.000000
std,195.620477,0.069894,0.932270,NaN,1.242141,357.503014,2.389539,53.352947,14.980682


In [461]:
#checking unique values
print("Ecommerce Dataset Payment Methods:")
print(ecommerce_df["payment_method"].unique())

print("\nTransaction Dataset Payment Methods:")
print(transaction_df["payment_method"].unique())

Ecommerce Dataset Payment Methods:
<StringArray>
['Cash on Delivery', 'Credit Card', 'PayPal', 'Bank Transfer']
Length: 4, dtype: str

Transaction Dataset Payment Methods:
<StringArray>
['Credit Card', 'UPI', 'COD', 'PayPal', 'Debit Card', 'Wallet']
Length: 6, dtype: str


### Payment Method Standardization

The two datasets used different names for the same payment method. In the transaction dataset COD was standardized to Cash on Delivery to maintain consistency across both datasets. Other payment methods were left unchanged because they represent distinct payment options.

In [462]:
transaction_df["payment_method"] = transaction_df["payment_method"].replace({"COD": "Cash on Delivery"})

# Verify the changes
print(transaction_df["payment_method"].unique())

<StringArray>
['Credit Card', 'UPI', 'Cash on Delivery', 'PayPal', 'Debit Card', 'Wallet']
Length: 6, dtype: str


In [463]:
print(ecommerce_df["category"].unique())
print(transaction_df["category"].unique())

<StringArray>
['Technology', 'Office Supplies', 'Furniture', 'Clothing & Accessories']
Length: 4, dtype: str
<StringArray>
['Home', 'Grocery', 'Electronics', 'Beauty', 'Fashion', 'Toys', 'Sports']
Length: 7, dtype: str


### Product Category Comparison observations

The two datasets use different product category classifications. Since they represent different sources and product taxonomies, the category names were not standardized. Each dataset's categories will be analyzed independently while comparing overall sales and transaction patterns.

## Data Validation

The numeric columns in both datasets were checked for unexpected negative values to identify any potential data quality issues before analysis.

In [464]:
# Checking for negative values in key numeric columns 
(ecommerce_df[["quantity", "price", "discount",
            "total_amount", "shipping_cost", "profit"]] < 0).sum()

quantity           0
price              0
discount           0
total_amount       0
shipping_cost      0
profit           272
dtype: int64

In [465]:
(transaction_df[["price", "discount", "quantity","total_amount", "shipping_cost","profit_margin"]] < 0).sum()

price               0
discount            0
quantity            0
total_amount        0
shipping_cost       0
profit_margin    6104
dtype: int64

### Observations

- No negative values were found in quantity, price, discount, total sales, or shipping cost.
- The Global E-Commerce dataset contains **272** records with negative profit values.
- The E-Commerce Sales Transactions dataset contains **6,104** records with negative profit margins.
- These values represent loss-making transactions rather than data quality issues, so they were retained for profitability analysis.

### Data Cleaning Observations

- Both datasets contain complete records with no missing values.
- Duplicate records were checked, and duplicates were removed if found.
- The `order_date` column was converted to the datetime format in both datasets.
- Payment method names were standardized by replacing **"COD"** with **"Cash on Delivery"** in the transaction dataset.
- Product category names were not standardized because the two datasets use different category classifications.
- No unexpected negative values were found in quantity, price, discount, total sales or shipping cost.
- The Global E-Commerce dataset contains **272** records with negative profit values and the E-Commerce Sales Transactions dataset contains **6,104** records with negative profit margins. These represent loss-making transactions and were retained for further profitability analysis.
- Both datasets are clean, consistent and ready for feature engineering, relational database design and SQL analysis.

# Data Transformation

### Creating New Features

In this section, new features are created from the existing data to support business analysis, SQL queries and visualizations.

### create date feature with manual code
ecommerce_df["order_year"] = ecommerce_df["order_date"].dt.year
ecommerce_df["order_month"] = ecommerce_df["order_date"].dt.month_name()
ecommerce_df["order_day"] = ecommerce_df["order_date"].dt.day

transaction_df["order_year"] = transaction_df["order_date"].dt.year
transaction_df["order_month"] = transaction_df["order_date"].dt.month_name()
transaction_df["order_day"] = transaction_df["order_date"].dt.day


In [466]:
def create_date_features(df):
    df["order_year"] = df["order_date"].dt.year
    df["order_month"] = df["order_date"].dt.month_name()
    df["order_day"] = df["order_date"].dt.day
    
    return df

ecommerce_df_df = create_date_features(ecommerce_df)
transaction_df = create_date_features(transaction_df)

ecommerce_df.head()

,order_id,order_date,customer_name,customer_segment,country,region,category,product_name,quantity,price,discount,total_amount,shipping_cost,profit,payment_method,order_year,order_month,order_day
0,ORD-11121,2023-01-02,Karen Suzuki,Corporate,United States,North America,Technology,Wireless Bluetooth Headphones,3,99.43,0,298.29,9.31,124.92,Cash on Delivery,2023,January,2
1,ORD-11244,2023-01-02,John Johansson,Corporate,Spain,Europe,Technology,Mechanical Gaming Keyboard,4,97.93,20,313.38,14.31,83.62,Cash on Delivery,2023,January,2
2,ORD-10325,2023-01-03,Jessica Garcia,Consumer,Mexico,North America,Office Supplies,Binder Clips Assorted 48pc,2,10.74,0,21.48,8.12,3.69,Credit Card,2023,January,3
3,ORD-10467,2023-01-03,Clara Taylor,Corporate,Italy,Europe,Technology,Webcam HD 1080p,2,61.86,15,105.16,10.79,26.32,PayPal,2023,January,3
4,ORD-11454,2023-01-05,Felix Thomas,Consumer,Italy,Europe,Furniture,Standing Desk Converter,4,330.67,20,1058.14,11.09,253.44,Credit Card,2023,January,5


In [467]:

transaction_df.head()

,order_id,customer_id,product_id,category,price,discount,quantity,payment_method,order_date,delivery_time_days,region,returned,total_amount,shipping_cost,profit_margin,customer_age,customer_gender,order_year,order_month,order_day
0,O100000,C17270,P234890,Home,164.08,0.15,1,Credit Card,2023-12-23,4,West,No,139.47,7.88,31.17,60,Female,2023,December,23
1,O100001,C17603,P228204,Grocery,24.73,0.00,1,Credit Card,2025-04-03,6,South,No,24.73,4.60,-2.62,37,Male,2025,April,3
2,O100002,C10860,P213892,Electronics,175.58,0.05,1,Credit Card,2024-10-08,4,North,No,166.80,6.58,13.44,34,Male,2024,October,8
3,O100003,C15390,P208689,Electronics,63.67,0.00,1,UPI,2024-09-14,6,South,No,63.67,5.50,2.14,21,Female,2024,September,14
4,O100004,C15226,P228063,Home,16.33,0.15,1,Cash on Delivery,2024-12-21,6,East,No,13.88,2.74,1.15,39,Male,2024,December,21


### Calculate Discount Amount

The two datasets store discount values in different formats.

- In the **Global E-Commerce Sales** dataset, the discount_percent column stores discounts as whole numbers so the values are divided by 100 before calculating the discount amount.
`discount_amount = unit_price × quantity × (discount_percent / 100)`

- In the **E-Commerce Sales Transactions** dataset, the discount column stores discounts as decimal values so no additional conversion is required.
`discount_amount = price × quantity × discount`

The discount amount is calculated by multiplying the product price, quantity, and discount value using the appropriate format for each dataset.

In [468]:
def calculate_discount_amount(df, price_col, quantity_col, discount_col, percentage=False):
    
    if percentage:
        df["discount_amount"] = (
            df[price_col] *
            df[quantity_col] *
            df[discount_col] / 100
        )
    else:
        df["discount_amount"] = (
            df[price_col] *
            df[quantity_col] *
            df[discount_col]
        )
        
    return df

In [469]:
ecommerce_df = calculate_discount_amount(
    ecommerce_df,
    "price",
    "quantity",
    "discount",
    percentage=True
)

ecommerce_df.head()


,order_id,order_date,customer_name,customer_segment,country,region,category,product_name,quantity,price,discount,total_amount,shipping_cost,profit,payment_method,order_year,order_month,order_day,discount_amount
0,ORD-11121,2023-01-02,Karen Suzuki,Corporate,United States,North America,Technology,Wireless Bluetooth Headphones,3,99.43,0,298.29,9.31,124.92,Cash on Delivery,2023,January,2,0.000
1,ORD-11244,2023-01-02,John Johansson,Corporate,Spain,Europe,Technology,Mechanical Gaming Keyboard,4,97.93,20,313.38,14.31,83.62,Cash on Delivery,2023,January,2,78.344
2,ORD-10325,2023-01-03,Jessica Garcia,Consumer,Mexico,North America,Office Supplies,Binder Clips Assorted 48pc,2,10.74,0,21.48,8.12,3.69,Credit Card,2023,January,3,0.000
3,ORD-10467,2023-01-03,Clara Taylor,Corporate,Italy,Europe,Technology,Webcam HD 1080p,2,61.86,15,105.16,10.79,26.32,PayPal,2023,January,3,18.558
4,ORD-11454,2023-01-05,Felix Thomas,Consumer,Italy,Europe,Furniture,Standing Desk Converter,4,330.67,20,1058.14,11.09,253.44,Credit Card,2023,January,5,264.536


In [470]:
transaction_df = calculate_discount_amount(
    transaction_df,
    "price",
    "quantity",
    "discount",
    percentage=False
)

transaction_df.head()

,order_id,customer_id,product_id,category,price,discount,quantity,payment_method,order_date,delivery_time_days,...,returned,total_amount,shipping_cost,profit_margin,customer_age,customer_gender,order_year,order_month,order_day,discount_amount
0,O100000,C17270,P234890,Home,164.08,0.15,1,Credit Card,2023-12-23,4,...,No,139.47,7.88,31.17,60,Female,2023,December,23,24.6120
1,O100001,C17603,P228204,Grocery,24.73,0.00,1,Credit Card,2025-04-03,6,...,No,24.73,4.60,-2.62,37,Male,2025,April,3,0.0000
2,O100002,C10860,P213892,Electronics,175.58,0.05,1,Credit Card,2024-10-08,4,...,No,166.80,6.58,13.44,34,Male,2024,October,8,8.7790
3,O100003,C15390,P208689,Electronics,63.67,0.00,1,UPI,2024-09-14,6,...,No,63.67,5.50,2.14,21,Female,2024,September,14,0.0000
4,O100004,C15226,P228063,Home,16.33,0.15,1,Cash on Delivery,2024-12-21,6,...,No,13.88,2.74,1.15,39,Male,2024,December,21,2.4495


### Create Loss Flag
This helps identify loss-making transactions.

In [471]:
ecommerce_df["loss_flag"] = ecommerce_df["profit"] < 0
ecommerce_df.head()

,order_id,order_date,customer_name,customer_segment,country,region,category,product_name,quantity,price,discount,total_amount,shipping_cost,profit,payment_method,order_year,order_month,order_day,discount_amount,loss_flag
0,ORD-11121,2023-01-02,Karen Suzuki,Corporate,United States,North America,Technology,Wireless Bluetooth Headphones,3,99.43,0,298.29,9.31,124.92,Cash on Delivery,2023,January,2,0.000,False
1,ORD-11244,2023-01-02,John Johansson,Corporate,Spain,Europe,Technology,Mechanical Gaming Keyboard,4,97.93,20,313.38,14.31,83.62,Cash on Delivery,2023,January,2,78.344,False
2,ORD-10325,2023-01-03,Jessica Garcia,Consumer,Mexico,North America,Office Supplies,Binder Clips Assorted 48pc,2,10.74,0,21.48,8.12,3.69,Credit Card,2023,January,3,0.000,False
3,ORD-10467,2023-01-03,Clara Taylor,Corporate,Italy,Europe,Technology,Webcam HD 1080p,2,61.86,15,105.16,10.79,26.32,PayPal,2023,January,3,18.558,False
4,ORD-11454,2023-01-05,Felix Thomas,Consumer,Italy,Europe,Furniture,Standing Desk Converter,4,330.67,20,1058.14,11.09,253.44,Credit Card,2023,January,5,264.536,False


In [472]:
transaction_df["loss_flag"] = transaction_df["profit_margin"] < 0
transaction_df.head()

,order_id,customer_id,product_id,category,price,discount,quantity,payment_method,order_date,delivery_time_days,...,total_amount,shipping_cost,profit_margin,customer_age,customer_gender,order_year,order_month,order_day,discount_amount,loss_flag
0,O100000,C17270,P234890,Home,164.08,0.15,1,Credit Card,2023-12-23,4,...,139.47,7.88,31.17,60,Female,2023,December,23,24.6120,False
1,O100001,C17603,P228204,Grocery,24.73,0.00,1,Credit Card,2025-04-03,6,...,24.73,4.60,-2.62,37,Male,2025,April,3,0.0000,True
2,O100002,C10860,P213892,Electronics,175.58,0.05,1,Credit Card,2024-10-08,4,...,166.80,6.58,13.44,34,Male,2024,October,8,8.7790,False
3,O100003,C15390,P208689,Electronics,63.67,0.00,1,UPI,2024-09-14,6,...,63.67,5.50,2.14,21,Female,2024,September,14,0.0000,False
4,O100004,C15226,P228063,Home,16.33,0.15,1,Cash on Delivery,2024-12-21,6,...,13.88,2.74,1.15,39,Male,2024,December,21,2.4495,False



### Convert percentage to decimal
The `discount` column was also converted from percentage values (e.g., 20) to decimal values (e.g., 0.20) so that both datasets use the same format.

In [473]:

ecommerce_df["discount"] = ecommerce_df["discount"] / 100

In [474]:
comparison = pd.DataFrame({
    "ECommerce": pd.Series(ecommerce_df.columns),
    "Transaction": pd.Series(transaction_df.columns)
})

comparison

,ECommerce,Transaction
0,order_id,order_id
1,order_date,customer_id
2,customer_name,product_id
3,customer_segment,category
4,country,price
5,region,discount
6,category,quantity
7,product_name,payment_method
8,quantity,order_date
9,price,delivery_time_days


# SQLite Database

This section creates a SQLite database based on the Entity Relationship Diagram (ERD). The cleaned datasets are loaded into relational tables so they can be queried using SQL.

In [479]:
conn = sqlite3.connect('../data/ecommerce_analysis.db')

global_products_df = (ecommerce_df[["product_name", "category"]].drop_duplicates().reset_index(drop=True))

global_sales_df = (ecommerce_df[["order_id","product_name","order_date","customer_name","customer_segment","country",
"region","quantity","price","discount","total_amount","shipping_cost","profit","payment_method",]].drop_duplicates().reset_index(drop=True))


customers_df = (transaction_df[["customer_id","customer_age","customer_gender",]].drop_duplicates().reset_index(drop=True))

transaction_sales_df = transaction_df[["order_id","customer_id","product_id","order_date","category","price","discount",
"quantity","payment_method","delivery_time_days","region","returned","total_amount","shipping_cost","profit_margin",]].copy()

transaction_products_df = (transaction_df[["product_id","category",]].drop_duplicates().reset_index(drop=True))


global_products_df.to_sql('global_products', conn, if_exists='replace', index=False)
global_sales_df.to_sql('global_sales', conn, if_exists='replace', index=False)
customers_df.to_sql('customers', conn, if_exists='replace', index=False)
transaction_sales_df.to_sql('transaction_sales', conn, if_exists='replace', index=False)
transaction_products_df.to_sql('transaction_products', conn, if_exists='replace', index=False)
conn.commit()

# SQL Queries and Analysis

The following SQL queries were performed on the SQLite database to analyze sales performance, profitability, customer behavior and product performance. 

### Total Sales and Profit by Product Category

In [485]:
query1 = """
select
    category,
    count(*) as total_orders,
    round(sum(total_amount),2) as total_sales,
    round(sum(profit),2) as total_profit
from global_sales gs
join global_products gp
on gs.product_name = gp.product_name
group by category
order by total_sales desc;
""" 

category_sales = pd.read_sql(query1, conn)
category_sales

,category,total_orders,total_sales,total_profit
0,Furniture,507,256274.68,81171.57
1,Technology,567,139518.22,48268.65
2,Clothing & Accessories,413,69375.63,26112.94
3,Office Supplies,513,19390.81,3319.16


### Top 10 Countries by Sales

In [486]:
query2 = """
select
    country,
    round(sum(total_amount),2) as total_sales
from global_sales
group by country
order by total_sales desc
limit 10;
"""

country_sales = pd.read_sql(query2, conn)
country_sales

,country,total_sales
0,Mexico,47217.30
1,Canada,45326.55
2,United States,41332.53
3,Japan,30950.19
4,United Kingdom,30185.46
5,Germany,28989.73
6,China,28322.85
7,Italy,27484.38
8,France,26616.81
9,Australia,25287.72


### Average Profit Margin by Customer Gender

In [487]:
query3 = """
select
    customer_gender,
    round(avg(profit_margin),2) as average_profit_margin
from transaction_sales ts
join customers c
on ts.customer_id = c.customer_id
group by customer_gender;
"""

gender_profit = pd.read_sql(query3, conn)
gender_profit

,customer_gender,average_profit_margin
0,Female,28.23
1,Male,28.13
2,Other,28.70


### Average Delivery Time by Payment Method

In [488]:
query4 = """
select
    payment_method,
    round(avg(delivery_time_days),2) as average_delivery_days
from transaction_sales
group by payment_method
order by average_delivery_days;
"""

delivery_time = pd.read_sql(query4, conn)
delivery_time

,payment_method,average_delivery_days
0,PayPal,4.77
1,UPI,4.80
2,Debit Card,4.81
3,Cash on Delivery,4.82
4,Credit Card,4.83
5,Wallet,4.85


### Products with Above-Average Profit

In [489]:
query5 = """
select
    product_name,
    profit
from global_sales
where profit >
(
    select avg(profit)
    from global_sales
)
order by profit desc
limit 10;
"""

top_profit = pd.read_sql(query5, conn)
top_profit

,product_name,profit
0,Corner L-Shaped Desk,1373.63
1,Corner L-Shaped Desk,1370.08
2,Ergonomic Office Chair,1189.79
3,Mesh Back Task Chair,1116.45
4,Ergonomic Office Chair,1115.05
5,Ergonomic Office Chair,1054.26
6,Corner L-Shaped Desk,1017.49
7,Mesh Back Task Chair,992.84
8,Standing Desk Converter,874.37
9,Standing Desk Converter,860.28
